### ISSUE TITLE
Jack and Jill generated accessor method names start with "-" which is incompatible with the jimple parser

### ISSUE THREAD

**Issue by agorski3** [soot-oss/soot#650]
Starting with Android 6 Google switched to using their own java compiler called "Jack and Jill" to compile java code directly into dex, skipping the previous intermediary compilation step in which java code was compiled into .class files before compiling these into dex.

Apparently when writing the "Jack and Jill" compiler, Google decided to implement their own naming scheme for the generated accessors to private fields of outer classes as well as other generated accessor methods. These generated accessor methods are named something to the effect of "access$000" when using the javac compiler. However, when using the "Jack and Jill" compiler, they appear as methods with names such as "-get0", "-wrap1", "-java_lang_CharSequence_lambda$1", or "-set0". Note I attached the file 
[methods_accessor_boot.oat_core-oj.txt](https://github.com/Sable/soot/files/517758/methods_accessor_boot.oat_core-oj.txt) which has some more examples of the generated method names created by "Jack and Jill".

As you might have guessed, these generated methods whose names begin with a dash can be a bit of an issue as method names beginning with a dash is non-standard (I believe). Most visibly, these method names are resulting in an exception in the Jimple parser when it tries to read back in jimple code generated from dex code compiled with "Jack and Jill". Here are a few of the exceptions I have seen and the code the parser was trying to parse when the exception occurred:

Exception 1:

java.lang.RuntimeException: Error parsing class java.lang.Class [1616,80] expecting: quoted name, identifier
    at soot.JimpleClassSource.resolve(JimpleClassSource.java:62)
    at soot.SootResolver.bringToHierarchy(SootResolver.java:243)
    at soot.SootResolver.bringToSignatures(SootResolver.java:275)
    at soot.SootResolver.bringToBodies(SootResolver.java:313)
    at soot.SootResolver.processResolveWorklist(SootResolver.java:166)
    at soot.SootResolver.resolveClass(SootResolver.java:134)
    at soot.Scene.tryLoadClass(Scene.java:774)
    at soot.Scene.loadBasicClasses(Scene.java:1434)
    at soot.Scene.loadNecessaryClasses(Scene.java:1513)
    at org.sag.pminer.sootinit.DexSootLoader.load(DexSootLoader.java:41)
    at org.sag.pminer.sootinit.AppDexSootLoader.load(AppDexSootLoader.java:39)
    at org.sag.sje.soot.DexToJimpleDecompiler.appsToJimple(DexToJimpleDecompiler.java:353)
    at org.sag.sje.soot.DexToJimpleDecompiler.run(DexToJimpleDecompiler.java:64)
    at org.sag.sje.JJExtractor.run(JJExtractor.java:56)
    at org.sag.sje.JJExtractor.main(JJExtractor.java:215)
Caused by: soot.jimple.parser.parser.ParserException: [1616,80] expecting: quoted name, identifier
    at soot.jimple.parser.parser.Parser.parse(Parser.java:1455)
    at soot.jimple.parser.JimpleAST.<init>(JimpleAST.java:57)
    at soot.JimpleClassSource.resolve(JimpleClassSource.java:43)
    ... 14 more

Code 1:

```
$r2 = staticinvoke <java.lang.Class$Caches: libcore.util.BasicLruCache -get0()>();
```

Jimple File 1:

[Class.jimple](https://github.com/Sable/soot/files/517760/Class_6_0_1_r1.txt)

Exception 2:

java.lang.RuntimeException: Error parsing class java.io.ObjectOutputStream [16,61] expecting: quoted name, identifier
    at soot.JimpleClassSource.resolve(JimpleClassSource.java:62)
    at soot.SootResolver.bringToHierarchy(SootResolver.java:243)
    at soot.SootResolver.bringToSignatures(SootResolver.java:275)
    at soot.SootResolver.bringToBodies(SootResolver.java:313)
    at soot.SootResolver.processResolveWorklist(SootResolver.java:166)
    at soot.SootResolver.resolveClass(SootResolver.java:134)
    at soot.Scene.tryLoadClass(Scene.java:774)
    at soot.Scene.loadBasicClasses(Scene.java:1434)
    at soot.Scene.loadNecessaryClasses(Scene.java:1513)
    at org.sag.pminer.sootinit.DexSootLoader.load(DexSootLoader.java:67)
    at org.sag.pminer.sootinit.AppDexSootLoader.load(AppDexSootLoader.java:39)
    at org.sag.sje.soot.DexToJimpleDecompiler.appsToJimple(DexToJimpleDecompiler.java:560)
    at org.sag.sje.soot.DexToJimpleDecompiler.run(DexToJimpleDecompiler.java:103)
    at org.sag.sje.JJExtractor.run(JJExtractor.java:78)
    at org.sag.sje.JJExtractor.main(JJExtractor.java:187)
Caused by: soot.jimple.parser.parser.ParserException: [16,61] expecting: quoted name, identifier
    at soot.jimple.parser.parser.Parser.parse(Parser.java:1455)
    at soot.jimple.parser.JimpleAST.<init>(JimpleAST.java:57)
    at soot.JimpleClassSource.resolve(JimpleClassSource.java:43)
    ... 14 more

Code 2:

```
static java.io.ObjectOutputStream$BlockDataOutputStream -get0(java.io.ObjectOutputStream)
{
```

Jimple File 2:

[ObjectOutputStream.jimple](https://github.com/Sable/soot/files/517762/ObjectOutputStream_7_0_0_r6.txt)

Exception 3:

java.lang.RuntimeException: Error parsing class java.util.Locale [43,26] expecting: quoted name, identifier
    at soot.JimpleClassSource.resolve(JimpleClassSource.java:62)
    at soot.SootResolver.bringToHierarchy(SootResolver.java:243)
    at soot.SootResolver.bringToSignatures(SootResolver.java:275)
    at soot.SootResolver.bringToBodies(SootResolver.java:313)
    at soot.SootResolver.processResolveWorklist(SootResolver.java:166)
    at soot.SootResolver.resolveClass(SootResolver.java:134)
    at soot.Scene.tryLoadClass(Scene.java:774)
    at soot.Scene.loadBasicClasses(Scene.java:1434)
    at soot.Scene.loadNecessaryClasses(Scene.java:1513)
    at org.sag.pminer.sootinit.DexSootLoader.load(DexSootLoader.java:67)
    at org.sag.pminer.sootinit.AppDexSootLoader.load(AppDexSootLoader.java:39)
    at org.sag.sje.soot.DexToJimpleDecompiler.appsToJimple(DexToJimpleDecompiler.java:559)
    at org.sag.sje.soot.DexToJimpleDecompiler.run(DexToJimpleDecompiler.java:102)
    at org.sag.sje.JJExtractor.run(JJExtractor.java:78)
    at org.sag.sje.JJExtractor.main(JJExtractor.java:185)
Caused by: soot.jimple.parser.parser.ParserException: [43,26] expecting: quoted name, identifier
    at soot.jimple.parser.parser.Parser.parse(Parser.java:1455)
    at soot.jimple.parser.JimpleAST.<init>(JimpleAST.java:57)
    at soot.JimpleClassSource.resolve(JimpleClassSource.java:43)
    ... 14 more

Code 3:

```
static java.util.Map -get0(java.util.Locale)
{
```

Jimple File 3:

[Locale.jimple](https://github.com/Sable/soot/files/517763/Locale_6_0_1_r1.txt)

Note these exceptions and code examples are from the jimple code generated from the dex of the standard java libraries included in both the system images of android 6 and android 7 (i.e. core-oj.jar's dex which is in boot.oat in both version or android). 

Further note, I have verified that dashed method names do not exist in the dex of system images for android 5 (which is before "Jack and Jill" was introduced). I have also verified that dashed method names exist within boot.oat and the classes.dex file of core-oj.jar before it is compiled into boot.oat. Moreover, I have verified that compiling the standard java libraries included in android with javac does not produce dashed method names but instead includes the normal "access$000" format. I was able to determine this because some classes in Android 6 were being compiled with both javac and "Jack and Jill" as the build process had not been completely migrated to "Jack and Jill". Thus I am fairly certain that these dashed method names are a result of the "Jack and Jill" compiler.

Unfortunately, as google is insistent on pushing the use of its "Jack and Jill" compiler for everything Android (i.e. the Android system and in the near future all Android apps), these dashed method names are going to be a problem now and in the future. 

So I see two solutions for this. Feel free to modify or add on to my solutions.
1. Modify the parser to identifiers that begin with dashes.
2. Quote these dashed identifiers when the jimple files are written as this seems to be what the parser is expecting anyways.

Let me know if you need any more information about this problem.

**Comment by agorski3** [comment]
The issue arises when reading in the jimple generated from the dex of the core java libraries which are inside of core-oj.jar for android 6.0.1_r1 and android 7.0.0_r6. I have attached the offending jar here [core-oj.zip](https://github.com/Sable/soot/files/524746/core-oj.zip).

**Comment by agorski3** [comment]
Recording a few things here from our email discussion here so they can be found easier.

Steps to reproduce the problem:

1. load core-oj.jar in soot
2. dump all classes to jimple
3. reset soot and try to reload the classes from the dumped jimple files
4. receive a lovely parser error

Possible Solutions:

Solution 1: Escape the method name on output if it contained the offending "-" character in the front

1. Add a method named "getEscapedName" to SootMethod that returns the escaped method name if it contains a "-" otherwise it returns the same as "getName".
2. Edit the locations where "getName" is used in the writing of jimple and replace with "getEscapedName".
3. Edit the "getSubSignatureImpl" function to recognize when a method needs to be escaped and escape it.
4. Edit the "grabMethod" of Scene. I am sure there is something in there that needs to be fixed given it is using the signature.
5. Edit the constructors in SootMethod and SootMethodRef to strip out the escape characters if they receive a method name containing them. This way reading in the methods from jimple is hopefully a bit simpler but may not be if you are keeping a list of method names somewhere that uses a parser outside of that in grabMethod to get the method name from a signature.

Solution 2: Modify the method names when they are being read from the dex file

Since these are automatically generated methods by the compiler the method name changes would most likely go unnoticed. This would likely be a lot easier to implement as the changes would be contained to the dexpler code. However, this is causing a change to the original code even if it is just to the compiler generated methods. I am not sure what your opinion is on such an action.

**Comment by agorski3** [comment]
A few more things you should be aware of as apparently the dash issue is not limited to jack and jill generated method names.

1. Jack and Jill generated instance and static fields also start with dashes and will most likely cause issues for the jimple parser. 

2. Jack and Jill generated class names also start with dashes. However, as these are all inner classes, they do not appear to be an issue for the jimple parser. For example, consider the actual generated class name:

    ```
    ManageApplications$ApplicationsAdapter$-void_-com_android_settings_applications_ManageApplications$ApplicationsAdapter_lambda$1_com_android_settingslib_applications_ApplicationsState$AppFilter_finalFilterObj_java_util_Comparator_comparatorObj_LambdaImpl0
    ```

    Note how the dash does not occur in the front of the class name because it is an inner class. Also note this class name does not include its package which for reference is `com.android.settings.applications`.

3. While there is no issue with the jimple parser when parsing generated class names, there is most likely an issue with constructing the inner class hierarchy because the generated class naming convention in jack and jill is down right awful. For example, the actual name of the class in the previous example minus the names of all its outer classes is:

    ```
    -void_-com_android_settings_applications_ManageApplications$ApplicationsAdapter_lambda$1_com_android_settingslib_applications_ApplicationsState$AppFilter_finalFilterObj_java_util_Comparator_comparatorObj_LambdaImpl0
    ```

    Note how the class name still contains multiple '$' which traditionally denote the inner class hierarchy. Also note that the class name of a generated inner class always begins with a dash and that everything after and including the dash is part of the generated class name. The dash essentially signifies that everything after it is to be part of the class name. However, since soot is not aware of this convention, situations like the following arise:

    ```
    Warning: com.android.settings.applications.ManageApplications$ApplicationsAdapter$-void_-com_android_settings_applications_ManageApplications$ApplicationsAdapter_lambda is a phantom
    ```
    
    Essentially, soot is searching for a non-existent outer class because it does not know how to read the jack and jill generated class names. This is just one small side effect of jack and jill awful class naming conventions that is not too big of a deal. There are most likely others however that are more detrimental.

**Comment by StevenArzt** [comment]
I changed the escaping to check whether an identifier starts with a dash (commit e8cc45f7df3e1dc73f3f19191917301a9c0e6e12). If so, the identifier is put in quotes. This re-uses the existing infrastructure that we already have for escaping invalid names in Jimple. In my tests, I was able to convert an APK file compiled with the Jack compiler to Jimple and from there back to a dex file. I'll next test your zip file as well.

**Comment by StevenArzt** [comment]
On the weird method names: They only get generated for lambda expressions, not for "normal" anonynous inner classes. Furthermore, these auto-generated lambda classes have no this$0 reference to their outer class. Still, there is are the two annotations:

Ldalvik/annotation/EnclosingClass
Ldalvik/annotation/InnerClass

This means that it is still an inner class, albeit non-standard. As long as you stay inside the Soot scne, this is all file. The naming convention is only used in the absence of these annotations. if you write out Jimple files, annotations are, however, lost. Jimple simply doesn't support tags and we don't have any other way to encode the annotations. That's a known limitation of the Jimple language. When you read the Jimple files back in, Soot needs to fall back to the naming convention to create the links between inner and outer class. Is that a major issue for you? It only happens if you go Scene -> Jimple files -> Scene, but not if you directly operate on the scene.

**Comment by agorski3** [comment]
Really, from what I can tell methods like "-get0" were replacing the access$000 methods of normal java.  But yes I understand methods like "-java_lang_CharSequence_lambda$1" are lambda expression generated methods. 

And in regards to your question, what would happen if I went Scene -> Jimple files -> Scene? Would there be some kind of parsing error with the jimple parser or is there some other issue? I am not sure what annotations are in this context.

**Comment by StevenArzt** [comment]
If you go Scene -> Jimple Files -> Scene, the auto-generated classes that violate the normal Java naming scheme will be loaded, but if you ask Soot for the outer class of such a class, you will likely get a phantom class of the wrong name, because Soot follows the normal Java naming scheme. The effect is what you described as point 3 above. Aside from that, the import should work.

**Comment by agorski3** [comment]
Hmm could we not just modify the code slightly so that when it is parsing the class name looking for '$', and it encounters a inner class name beginning with a '-', then it will treat the entire string after and including the '-' and any additional '$' as the class name. Would that not fix issue 3?

**Comment by agorski3** [comment]
Turns out the warnings I was seeing were actually because dexpler's procedure for determining outer classes was failing when it encountered the jack and jill generated lambda class names. I fixed this by adding a special case to the parsing procedure. I also fixed some issues with the parsing procedure for getting the outer class of an inner class in general (i.e. inconsistent regular expressions that sometimes matched too much). There is more information on this in the pull request at
https://github.com/Sable/soot/pull/657.

I also discovered that the procedure for reading in jimple classes never constructs the outer/inner class hierarchy. I added a small amount of code to perform this construction as the jimple source is being read in. This code is also in the pull request mentioned.
